### PR TITLE
Mdd on trades backtest stat

### PR DESCRIPTION
**Pull request by Guitheg**
## Summary

Implementation of the MDD on the trades intervals

Solve the half of this issue: ##5106

## Quick changelog

# Add
- new function calculate_trades_mdd
- visualization of the result of the new function in the stats board at the end of a backtest
- unit test of the function

## What's new?
Added mdd on trades stat on the stats board at the end of backtest
image

### CODE REVIEW

**Comment by xmatthias** [comment]
Something seems broken with this - but i'm not sure what (i suspect it's the deletion of the dataframe in the function, but not 100% sure).

develop 
```
=============== SUMMARY METRICS ================
| Metric                 | Value               |
|------------------------+---------------------|
| Backtesting from       | 2021-01-01 00:00:00 |
| Backtesting to         | 2021-10-08 16:00:00 |
| Max open trades        | 2                   |
|                        |                     |
| Total/Daily Avg Trades | 429 / 1.53          |
| Starting balance       | 1000.000 USDT       |
| Final balance          | 987.497 USDT        |
| Absolute profit        | -12.503 USDT        |
| Total profit %         | -1.25%              |
| Trades per day         | 1.53                |
| Avg. daily profit %    | -0.00%              |
| Avg. stake amount      | 20.000 USDT         |
| Total trade volume     | 8580.000 USDT       |
|                        |                     |
| Best Pair              | MATIC/USDT 33.03%   |
| Worst Pair             | XTZ/USDT -39.20%    |
| Best trade             | WAVES/USDT 26.55%   |
| Worst trade            | XMR/USDT -3.89%     |
| Best day               | 7.170 USDT          |
| Worst day              | -3.896 USDT         |
| Days win/draw/lose     | 87 / 82 / 112       |
| Avg. Duration Winners  | 17:04:00            |
| Avg. Duration Loser    | 7:02:00             |
| Rejected Buy signals   | 41436               |
|                        |                     |
| Min balance            | 977.748 USDT        |
| Max balance            | 1007.028 USDT       |
| Drawdown               | 126.17%             |
| Drawdown               | 25.259 USDT         |
| Drawdown high          | 7.028 USDT          |
| Drawdown low           | -18.231 USDT        |
| Drawdown Start         | 2021-05-18 16:00:00 |
| Drawdown End           | 2021-07-21 20:00:00 |
| Market change          | 676.00%             |
================================================
```

This branch
```
=================== SUMMARY METRICS ====================                                                                           
| Metric                   | Value                     |                                                                           |--------------------------+---------------------------|                                                                           
| Backtesting from         | 2021-01-01 00:00:00       |                                                                           
| Backtesting to           | 2021-10-08 16:00:00       |                                                                           
| Max open trades          | 2                         |                                                                           
|                          |                           |                                                                           
| Total/Daily Avg Trades   | 429 / 1.53                |                                                                           
| Starting balance         | 1000.000 USDT             |                                                                           
| Final balance            | 987.497 USDT              |                                                                           
| Absolute profit          | -12.503 USDT              |                                                                           
| Total profit %           | -1.25%                    |                                                                           
| Trades per day           | 1.53                      |                                                                           
| Avg. daily profit %      | -0.00%                    |                                                                           
| Avg. stake amount        | 20.000 USDT               |                                                                           
| Total trade volume       | 8580.000 USDT             |                                                                           
|                          |                           |                                                                           
| Best Pair                | MATIC/USDT 33.03%         |                                                                           
| Worst Pair               | XTZ/USDT -39.20%          |                                                                           
| Best trade               | WAVES/USDT 26.55%         |                                                                           
| Worst trade              | XMR/USDT -3.89%           |                                                                           
| Best day                 | 7.170 USDT                |                                                                           
| Worst day                | -3.896 USDT               |                                                                           
| Days win/draw/lose       | 87 / 82 / 112             |                                                                           
| Avg. Duration Winners    | 17:04:00                  |                                                                           
| Avg. Duration Loser      | 7:02:00                   |                                                                           
| Rejected Buy signals     | 41436                     |                                                                           
|                          |                           |                                                                           
| Min balance              | 0.000 USDT                |         <-- Everything from here to the bottom is wrong except for Market change.
| Max balance              | 0.000 USDT                |                                                                           
|                          |                           |         
| Max Drawdown (in trades) | 0.00%                     |                                                                           
| Max Drawdown Total       | 0.00%                     |         
| Max Drawdown Total       | 0.000 USDT                |                                                                           
| Drawdown high            | 0.000 USDT                |         
| Drawdown low             | 0.000 USDT                |                                                                           
| Drawdown Start           | 1970-01-01 00:00:00+00:00 |                                                                           
| Drawdown End             | 1970-01-01 00:00:00+00:00 |                                                                           
| Market change            | 676.00%                   |                                                                           
======================================================== 
```


As you can see, all drawdown (as well as min/max) balances) are set to 0 for no apparent reason, which indicates a bug "someplace".

---

apart from that (i can't test it myself due to the above at the moment) - from a first look, it looks like you're calculating the maximum drawdown "per pair" - and then use the maximum of that (basically showing the biggest drawdown for a single pair).
I don't think that's what MDD is supposed to show - as it simply ignores wins from all other pairs during the same period.

Assume i trade 3 pairs - XRP/USDT, BTC/USDT and ETH/USDT for the span of a month - with a max-open-trades of 1 (so i go all in on each pair when a trade is opened).
All trades are bought in sequence (always XRP/BTC/ETH - in the same sequence) - for a few rounds (say 5 rounds, which will give me 3x5 = 15 trades) - and each stop / take profit is at 10% (so all trades are even).

I lose all XRP trades - but win all BTC and ETH trades ... (so the resuts are 5x -10% and  10x + 10%).

The way i understand this implementation, then this would show the max-drawdown for the XRP trades (5 losses, resulting in 50% loss on XRP) - but completely ignores the BTC and ETH wins for 100%. 
In summary, the account would be "roughly" 50% up (depends if we use compounding or not)


based on the [investopedia article](https://www.investopedia.com/terms/m/maximum-drawdown-mdd.asp)  - this should be calculated against the portfolio - not using the maximum of single pairs.

**Comment by Guitheg** [comment]
I'm still investigate to know where it broke. I think the function sent a ValueError due to a bad using of isinstance used to check the dataframe.

About the implementation, you understood well what we did with @EmbraysiteReal. We based our implementation about what we read and what we understood. In fact, what we did looks like the investopedia method. But I think the explanation gave by investopedia is better considering the robustness and the generalization of the method. With @EmbraysiteReal we propose this definition of the max drawdown based on investopedia method. we formalize it mathematically to avoid ambiguity. (however tell us if you need clarification)
image

I have a question about the current max drawdown. The current max drawdown implemented on freqtrade consider only profit (so at a trade closing) to perfrom its calculation. In other words, the date k (defined in the equations) is updated only when a trade is closing. Then I think the drawdown did not record well all the evolution of the portfolio. Am I right ? Or there a something I did not understand ?

**Comment by xmatthias** [comment]
If you're looking for the "current outstanding" balance, maybe look at #3059 before trying to implement it directly / again.

In my opinion, that PR is "mostly" complete (at least for that function) - the only thing that I'm missing there really is a "proper" usecase (i don't like adding code just for the fun of it - and i don't think the loss function that's shown there  as sample usecase makes much sense).

based on that - i think you could base your drawdown calculation on the outcome of that function - as that would be the "real" max drawdown, including non-realized profits.

**Comment by Guitheg** [comment]
so, instead of adding the `calculate_trades_mdd` function. You propose to just rewrite the calculation of the current `calculate_max_drawdown` function following the definition we previously gave ?

We will begin to answer #3059. Indeed it would be useful to have such a function in order to calculate the max drawdown.

**Comment by xmatthias** [comment]
neither - `calculate_max_drawdown` works perfectly fine and according to spec - but results will vary depending on the input / selected column.

In my understanding, the investopedia definition can be archived pretty easily by using `calculate_max_drawdown` against absolute profits (we do this for the absolut profit calculation already). 
To then get the "relative" value - you'd calculate the % difference between the returned high and low value.

That said, it'll obviously not include non-realized losses - which can however be accomplished by first running the data through the "outstanding profits" function.

**Comment by EmbraysiteReal** [comment]
Hello, could you send to us the spec that you used for `calculate_max_drawdown` ? We don't understand what it represent since we only know the investopedia MDD definition.

**Comment by xmatthias** [comment]
It's a very simple, drawdown calculation.

it's searching the highest point in a curve - followed by the subsequent deepest valley.
If you use `freqtrade plot-profit` you'll see the highest / lowest points - which will make the way of calculation clear.

**Comment by EmbraysiteReal** [comment]
It's weird, why is there an issue to implement MDD, which mean max drawdown, if `calculate_max_drawdown` already works perfectly ?
And i don't know if you saw result like i did, but sometimes the drawdown can be superior than 100%, this is really weird and completely out of meaning for me, maybe i broke something

**Comment by xmatthias** [comment]
Let me repeat - `calculate_max_drawdown` works well (from a function perspective) - but depends on the input given (you can see this with absolute drawdowns, which you'll notice if you compare it to the plot - is accurate).

Relative drawdown (like -120%) which is currently used are not wrong, just need some context for correct interpretation.
This is discussed in parts in #5106, and in more depth in #5167.
Nobody said the calculation per se is wrong (you assumed this without checking how the calculation actually worked) - just that what is shown is confusing (it's also only wrong in certain conditions - see below).

You can even test that pretty easily with a random series (in a jupyter notebook) - and plot the results as well as high/low points.

First of all, it does only consider finished / profits. 
When using a sane stoploss (10% or less) - this is not a problem.
With strategies that use a very wide or no stoploss (99%) - this will make drawdown seem great - while in reality it might not become a problem - but again, it's not something i'd see as critical.

The bigger problem in my opinion is the aggregation of relative profits.
When the metric was developed - only static stake-amounts were allowed - where it works perfectly fine (10% profit is always 10% per `stake_amount`).
With the introduction of "unlimited" stake amount, profits can compound (which is good in general for profits - usually).

However, you can no longer assume that 10% + -10% = 0%.
it could be 10% * 50$ - and -10% * 2000$ - so now we're -1950$ - BUT - relative aggregation still shows 0% profits.

THIS is the real problem with (relative) drawdown calculation. It uses the relative cumulative profit - while it really should not.
in my opinion, it can be pretty easily fixed by using the absolute values (drawdown high value+drawdown low value) we already show as metrics - and simply calculate the relative difference between the two (so from a high 2000$ to a low of 500$ - you've had a drawdown of 75%.


Now the first point - using "intermediate" profits - might be seen as separate point to tackle - but again, won't require a new calculating function - but simple data preparation (basically what is implemented in #3059 - maybe with some modifications) before passing the data into the drawdown calculation function.

**Comment by EmbraysiteReal** [comment]
Thank you a lot for your answer, there is a lot of what you'r saying that we saw. It clarify a lot of things, the aggregation of relative profits was an aberration from my mathematician point of view

"Nobody said the calculation per se is wrong (you assumed this without checking how the calculation actually worked)" we checked the calculation and we saw for exemple the aggregation of relative profits, wich is a wrong way of calculation, then we tried to understood why it was done this way then we assume they were something wrong. I don't even understand what you mean by "calculation per se".

We are new to the project and we don't want to pretend that we know more than others, but when we found somethings we don't understand in the code we can't ignore it like nothing happened. There is no documentation and comments on the code are minimalist, it's normal to have misunderstanding

Well you're right, we have to check #3059 since it's a big part of the solution we proposed. We will try to make it optimized as possible.
And again, thanks for the clear answer

**Comment by xmatthias** [comment]
> I don't even understand what you mean by "calculation per se".

the calculation of the method `calculate_max_drawdown` is correct depending on the inputs. If you use `value_col='profit_abs'` as parameter (as we do for absolute profits) - then the results are correct.
they are also technically correct for `value_col='profit_ratio'` - but ONLY if stake-amount is static.

> There is no documentation and comments on the code are minimalist

Code is usually self-documenting.
it just requires some understanding of pandas (how pandas functions work, what they do/return) - and carefully thinking through the method - eventually multiple times.

Even a comment would only tell you `calculate cumsum over value_col` - if you don't understand what a cumulative sum is, that won't help you. If you do - the code already told you what you needed to know, so the comment is pointless.

--

We do have quite extensive user documentation.
I'll however not start documenting internal functions more than they already are.

They mostly have a doc-string describing what they do, as well as the parameters defined (what is expected).

The implementation itself is 
a) subject to change in case of bugs
b) clear from the code

**Comment by xmatthias** [comment]
Closing this in favor of #6165 - which uses the existing (and working) drawdown calculation to calculate account drawdown.
